crates/matrix-sdk/src/notification_settings/rules.rs

```diff
@@ -0,0 +1,1091 @@
+        } else if let Some(rule) = self.ruleset.get(kind, rule_id) {
+            Ok(rule.enabled())
+        } else {
+            Err(NotificationSettingsError::RuleNotFound)
+        }
+    }
+
+    /// Insert a new `Room` push rule for a given `room_id` and return an
+    /// optional `Command` describing the action to be performed on the
+    /// user's account data.
+    pub(crate) fn insert_room_rule(
+        &mut self,
+        kind: RuleKind,
+        room_id: &RoomId,
+        notify: bool,
+    ) -> Result<Option<Command>, NotificationSettingsError> {
+        let command;
+        let actions = if notify {
+            vec![Action::Notify, Action::SetTweak(Tweak::Sound("default".into()))]
+        } else {
+            vec![]
+        };
+
+        match kind {
+            RuleKind::Override => {
+                // Insert a new push rule matching this `room_id`
+                let new_rule = NewConditionalPushRule::new(
+                    room_id.to_string(),
+                    vec![PushCondition::EventMatch {
+                        key: "room_id".into(),
+                        pattern: room_id.to_string(),
+                    }],
+                    actions,
+                );
+                let new_rule = NewPushRule::Override(new_rule);
+                self.ruleset.insert(new_rule.clone(), None, None)?;
+                command = Some(Command::SetPushRule { scope: RuleScope::Global, rule: new_rule });
+            }
+            RuleKind::Room => {
+                // Insert a new `Room` push rule for this `room_id`
+                let new_rule = NewSimplePushRule::new(room_id.to_owned(), actions);
+                let new_rule = NewPushRule::Room(new_rule);
+                self.ruleset.insert(new_rule.clone(), None, None)?;
+                command = Some(Command::SetPushRule { scope: RuleScope::Global, rule: new_rule });
+            }
+            _ => {
+                return Err(NotificationSettingsError::InvalidParameter(
+                    "kind must be either Override or Room.".to_owned(),
+                ))
+            }
+        }
+
+        Ok(command)
+    }
+
+    /// Deletes a list of rules and returns a list of `Command` describing the
+    /// actions to be performed on the user's account data.
+    ///
+    /// # Arguments
+    ///
+    /// * `rules` - A list of rules to delete
+    /// * `exceptions` - A list of rules to ignore
+    pub(crate) fn delete_rules(
+        &mut self,
+        rules: &[(RuleKind, String)],
+        exceptions: &[(RuleKind, String)],
+    ) -> Result<Vec<Command>, RemovePushRuleError> {
+        let mut commands: Vec<Command> = vec![];
+
+        // Remove exceptions from the rules to delete
+        let to_remove = BTreeSet::from_iter(exceptions);
+        let mut rules = rules.to_vec();
+        rules.retain(|e| !to_remove.contains(e));
+
+        for (rule_kind, rule_id) in rules {
+            self.ruleset.remove(rule_kind.clone(), rule_id.clone())?;
```

Comment:
Do we expect many rules to be in `exceptions`? I would imagine not, and if I'm not mistaken there's a strategy that involves 0 allocations (that is, get rid of both the `BTreeSet::from_iter` and the `.to_vec()` call):

- iterate on the rules
- at the top of the `for` loop, do a linear search of the current rule in the `exceptions` list, and `continue` if it's found
- otherwise, do the `ruleset.remove` / `commands.push` calls

Then we can skip the two allocations for the BTreeSet + vector.

Reply:
Ok thanks, I've updated this part.